### DISCUSSION TITLE
Github show verified badge on commits signed after key expiration date

### DISCUSSION THREAD

**Discussion by pietrushnic** [community/community#72294] [Code Security]
### Select Topic Area

Bug

### Body

Using gpg 2.2.40, I used an expired key to sign the git commit. Despite that, Github shows a verified badge for that commit, claiming that `This commit was signed with the committer’s verified signature. The key expired after the commit was signed.`.

https://github.com/pietrushnic/gh-gpg-testing-repo/commit/32a404f8c04442965eafc4b71d240c25a1d3fd60

![signed_w_expired_key](https://github.com/community/community/assets/1071894/5dd77b29-d4c1-4cb1-85aa-44054c6ba82e)

Here are some more details. `git log --show-signature`

```shell
commit 6ef3aa2e8186bafa168a1451e78346d018b20c4e (HEAD -> main, origin/main)
gpg: Signature made Thu 19 Oct 2023 01:07:59 AM CEST
gpg:                using RSA key 2BF8D9D82D872920449A79C9FD8B7B3145214C77
gpg: Good signature from "Piotr Król Signature After Expiration Date (Employee Cert Key) <piotr.krol@3mdeb.com>" [unknown]
gpg: Note: This key has expired!
Primary key fingerprint: 04C9 7C54 E51F EBA3 0806  D8FC 40BA BCE7 542B 0B4A
     Subkey fingerprint: 2BF8 D9D8 2D87 2920 449A  79C9 FD8B 7B31 4521 4C77
Author: Piotr Król <piotr.krol@3mdeb.com>
Date:   Thu Oct 19 01:07:59 2023 +0200

    README: sign after key expiration 2
```

```shell
[22:43:24] user:gh-gpg-testing-repo git:(main) $ gpg --edit-key 04C97C54E51FEBA30806D8FC40BABCE7542B0B4A 
gpg (GnuPG) 2.2.40; Copyright (C) 2022 g10 Code GmbH
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Secret subkeys are available.

pub  rsa4096/40BABCE7542B0B4A
     created: 2022-10-14  expires: never       usage: C   
     trust: unknown       validity: unknown
sub  rsa4096/FD8B7B3145214C77
     created: 2022-10-14  expired: 2023-10-14  usage: S   
sub  rsa4096/C579B798739ACC35
     created: 2022-10-14  expired: 2023-10-14  usage: E   
[ unknown] (1). Piotr Król Signature After Expiration Date (Employee Cert Key) <piotr.krol@3mdeb.com>

gpg> 
```

Please let me know if this is a bug or if there is something wrong with my above validation process.

**Comment by Finn-paulsen** [comment]
Response #1

**Reply by pietrushnic** [reply]
@Finn-paulsen IIUC, this is the incorrect place to report such an issue since this discussion is only for `GitHub Discussions (public beta), Mobile or Codespaces (limited public beta)` ? If so, what is the correct place to report issues with/the GH interface? I guess GitHub should consider to make it more obvious.

**Comment by deadnameremoval** [comment]
It's not an actual bug if the commit is known to be made while the signature was valid (i.e., the commit was already uploaded to github).

Is this the case? I'm not sure I understood.

**Reply by pietrushnic** [reply]
I reported the issue directly to GH, and they confirmed that the message could be confusing; they also escalated the problem to engineers; please check [this thread](https://github.com/3mdeb/3mdeb-secpack/pull/101#discussion_r1372886255) for the whole context.

In general, relying on the green badge status of GH is a mistake, as well as on comments regarding key expiration and revocation. Those who care should always verify the signatures themselves. The key piece of documentation is here:
https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-gpg-key-to-your-github-account

> To verify as many of your commits as possible, you can add expired and revoked keys. If the key meets all other verification requirements, commits that were previously signed by any of the corresponding private keys will show as verified and indicate that their signing key is expired or revoked.

Misleading is also `The key expired after the commit was signed` because we proved this information does not change even if we try to [time travel](https://github.com/3mdeb/3mdeb-secpack/pull/101#discussion_r1364674236). As my colleague wrote, this means that when your key gets stolen, even after its expiry date, someone can use it to sign a commit that will be green in GUI. Of course, another valid point in the discussion was that you really can't have any claims about time without involving an external authority that will sign a timestamp to prove the signature was made at that precise time since time can be faked pretty easily as it was shown in the linked thread.

**Comment by mukhriddin-dev** [comment]
response #6
